### PR TITLE
Add module-evm-accounts

### PR DESCRIPTION
**Pull request by zjb0807**
Closes: #478

### CODE REVIEW

**Comment by zjb0807** [comment]
Need more unit test and integration test

File: modules/evm-accounts/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+		fn deposit_event() = default;
+
+		#[weight = T::WeightInfo::claim_account()]
+		pub fn claim_account(origin, eth_signature: EcdsaSignature, eth_address: EvmAddress) {
+			with_transaction_result(|| {
+				let who = ensure_signed(origin)?;
+
+				// recover evm address from signature
+				let address = Self::eth_recover(&eth_signature, &who.using_encoded(to_ascii_hex), &[][..]).ok_or(Error::<T>::BadSignature)?;
+				ensure!(eth_address == address, Error::<T>::BadAddress);
```
Comment (xlc) [inline modules/evm-accounts/src/lib.rs:117]:
```suggestion
				ensure!(eth_address == address, Error::<T>::BadSignature);
```

or `InvalidSignature`

File: modules/evm-accounts/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+				// check if the evm padded address already exists
+				let account_id = T::AddressMapping::into_account_id(eth_address);
+				let mut nonce = <T as frame_system::Trait>::Index::default();
+				if frame_system::Account::<T>::contains_key(account_id.clone()) {
```
Comment (xlc) [inline modules/evm-accounts/src/lib.rs:122]:
```suggestion
				if frame_system::Module::<T>::is_explicit(&account_id) {
```

File: modules/evm-accounts/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+					T::Currency::unreserve(&account_id, total_reserved);
+
+					// transfer all free to origin
+					T::Currency::transfer(&account_id, &who, T::Currency::free_balance(&account_id), ExistenceRequirement::AllowDeath)?;
+
+					nonce = frame_system::Module::<T>::account_nonce(&account_id);
+					// finally kill the account
+					T::KillAccount::happened(&account_id);
+
+					EvmAddresses::<T>::remove(account_id.clone());
```
Comment (xlc) [inline modules/evm-accounts/src/lib.rs:151]:
```suggestion
					EvmAddresses::<T>::remove(&account_id);
```

File: modules/evm-accounts/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+				}
+				//	make the origin nonce the max between origin amd evm padded address
+				let origin_nonce = frame_system::Module::<T>::account_nonce(&who);
+				if origin_nonce < nonce {
+					for _ in 0..(origin_nonce - nonce).unique_saturated_into() {
```
Comment (xlc) [inline modules/evm-accounts/src/lib.rs:156]: should just use `frame_system::Account::<T>::mutate` to make it fast

File: modules/evm-accounts/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+						frame_system::Module::<T>::inc_account_nonce(&who);
+					}
+				}
+
+				// update accounts
+				Accounts::<T>::mutate(eth_address, |v| *v = who.clone());
+				EvmAddresses::<T>::insert(who.clone(), eth_address);
```
Comment (xlc) [inline modules/evm-accounts/src/lib.rs:163]:
```suggestion
				EvmAddresses::<T>::insert(&who, eth_address);
```

**Review by xlc — APPROVED** [review]

File: modules/evm-accounts/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+		fn deposit_event() = default;
+
+		#[weight = T::WeightInfo::claim_account()]
+		pub fn claim_account(origin, eth_signature: EcdsaSignature, eth_address: EvmAddress) {
```
Comment (xlc) [inline modules/evm-accounts/src/lib.rs:111]:
```suggestion
		pub fn claim_account(origin, eth_address: EvmAddress, eth_signature: EcdsaSignature) {
```

**Review by xlc — APPROVED** [review]

**Pull request merged**